auth/couchdb_session_authenticator_test.go

```diff
@@ -232,25 +233,36 @@ var _ = Describe("Authenticator Unit Tests", func() {
 
 			for i := 1; i <= 3; i++ {
 				wg.Add(1)
-				go func() {
+				go func(n int) {
 					defer GinkgoRecover()
 					defer wg.Done()
 					cookie, err := auth.getCookie()
-					Expect(err).To(BeNil())
-					Expect(cookie).To(Equal(oldCookie))
-					Expect(cookie.Value).To(Equal("fakefake-1"))
-				}()
+					// make sure that at least first refresh is async
+					// and returns an old still-valid cookie
+					if n == 1 {
+						Expect(err).To(BeNil())
+						Expect(cookie).To(Equal(oldCookie))
+						Expect(cookie.Value).To(Equal("fakefake-1"))
+					}
+				}(i)
 			}
 			wg.Wait()
 
-			// give refresher goroutine time to finish
-			time.Sleep(1 * time.Second)
+			// wait until server counter increased by refresher
+			Eventually(func() int {
+				return int(atomic.LoadInt32(&callNumber))
+			}).Should(Equal(2))
 
 			newCookie, err := auth.getCookie()
 			Expect(err).To(BeNil())
 			Expect(newCookie).ToNot(Equal(oldCookie))
```

Comment:
I saw this assertion fail when running locally with this branch - I think that means the removal of the 1 second wait has opened a new window where the session call increment has happened, but the parsing of the response cookie and passing of it to the refresh channel has not yet completed.

I suggest we just put this subsequent `getCookie()` assertion in an `Eventually` with a timeout, so we assert immediately if the refresh has fully completed or poll a few times if it hasn't.

Reply:
How you execute the tests that you find these cases? Do you use a bash loop or something else?

Reply:
Sure a loop, but it wasn't really necessary, timing is such on my machine that I never needed to get past 2 invocations to see a failure.

Reply:
With this bash script I could not catch the problem:
```bash
#!/usr/bin/env bash

counter=1
while [ $counter -le 10 ]
do
  go clean -testcache
  go test ./auth
  ((counter++))
done
echo All done
```
What is yours?

Reply:
fwiw I couldn't reproduce it even in ~10min loop, so we are lucky @ricellis has such a _special_ MBP, close by efficiency to a resource limited Jenkins 😛

yet this case should be fixed now with me calling it in `Eventually` with the latest change I pushed.

I'm going to merge this patch. It's not ideal, but hopefully will be stable enough to not cause the issues in CI

Reply:
🤞